rapids-cmake/cpm/thrust.cmake

```diff
@@ -0,0 +1,78 @@
+#=============================================================================
+# Copyright (c) 2020-2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+rapids_cpm_thrust
+-----------------
+
+.. versionadded:: v21.10.00
+
+Allow projects to find or build `Thrust` via `CPM` with built-in
+tracking of these dependencies for correct export support.
+
+Uses version 1.12.0 of Thrust for consistency across all RAPIDS projects
+
+.. code-block:: cmake
+
+  rapids_cpm_thrust( NAMESPACE <namespace>
+                     [BUILD_EXPORT_SET <export-name>]
+                     [INSTALL_EXPORT_SET <export-name>]
+                   )
+
+``NAMESPACE``
+  The namespace that the Thrust target will be constructed into.
+
+``BUILD_EXPORT_SET``
+  Record that a :cmake:command:`CPMFindPackage(<PackageName> ...)` call needs to occur as part of
+  our build directory export set.
+
+``INSTALL_EXPORT_SET``
+  Record a :cmake:command:`find_dependency(<PackageName> ...)` call needs to occur as part of
+  our build directory export set.
+
+.. note::
+  Installation of Thrust will occur if an INSTALL_EXPORT_SET is provided, and Thrust
+  is added to the project via :cmake:command:`add_subdirectory` by CPM.
+
+
+Result Targets
+^^^^^^^^^^^^^^
+  <namespace>::Thrust target will be created
+
+#]=======================================================================]
+function(rapids_cpm_thrust NAMESPACE namespaces_name)
+  list(APPEND CMAKE_MESSAGE_CONTEXT "rapids.cpm.thrust")
+
+  set(to_install OFF)
+  if(INSTALL_EXPORT_SET IN_LIST ARGN)
+    set(to_install ON)
+  endif()
+
+  include("${rapids-cmake-dir}/cpm/find.cmake")
+  rapids_cpm_find(Thrust 1.12.0 ${ARGN}
```

Comment:
It would be tricky to have a central config file to control the versions of non rapids branch naming projects, but it'd be useful to have a readme with a list of the dependencies and versions, particularly if we put more of them here, to be sure they are synced with integration, conda recipes, etc. What do you think?

Reply:
I agree

How about something like a json file in the cpm directory that listed all the project url and branch?

rapids-cmake and the docs could both reference it to maintain a single source of truth

Reply:
That sounds like a great solution!

Reply:
What I am thinking is something like:

```
{
  "GTest" : {
    "version" : "1.10",
    "git_url" : "https://github.com/google/googletest.git",
    "git_tag" : "release-{version}",
  },
  "nvbench" : {
    "version" : "0.0",
    "git_url" : "https://github.com/NVIDIA/nvbench.git",
    "git_tag" : "main",
  },
  "rmm" : {
    "version" : "rapids-cmake-version",
    "git_url" : "https://github.com/rapidsai/rmm.git",
    "git_tag" : "branch-{version}",
  },
  "spdlog" : {
    "version" : "1.8.5",
    "git_url" : "https://github.com/gabime/spdlog.git",
    "git_tag" : "v{version}",
  },
  "Thrust" : {
    "version" : "1.12.0",
    "git_url" : "https://github.com/NVIDIA/thrust.git",
    "git_tag" : "{version}",
  },
}

```

I am still trying to figure out how best to match `rmm` version to the current rapids calver, so for now I have used a placeholder ( rapids-cmake-version ).

Any thoughts?

Reply:
I have updated the PR with an initial implementation that has everything working based on a json file  ( https://github.com/rapidsai/rapids-cmake/pull/52/files#diff-695faffc4b8bc9bc2b2fb2f4c3958c279c2d01744eeead1a30dedd1ead81eb6c )

I still have to do the following:

- Reference the json versions in each packages documentation
- Formally document the versions.json file and what magic placeholders are supported